src/renderer/rendered_layer.rs

```diff
@@ -0,0 +1,608 @@
+        let has_transparency = default_background.a() != 255
+            || self.windows.iter().any(|window| window.has_transparency());
+
+        self._draw_shadow(root_canvas, &silhouette, settings);
+
+        root_canvas.save();
+        root_canvas.clip_path(&silhouette, None, Some(false));
+        let need_blur = has_transparency || settings.floating_blur;
+
+        if need_blur {
+            if let Some(blur) = blur(
+                (
+                    settings.floating_blur_amount_x,
+                    settings.floating_blur_amount_y,
+                ),
+                None,
+                None,
+                None,
+            ) {
+                let paint = Paint::default()
+                    .set_anti_alias(false)
+                    .set_blend_mode(BlendMode::Src)
+                    .to_owned();
+                let save_layer_rec = SaveLayerRec::default()
+                    .backdrop(&blur)
+                    .bounds(&bound_rect)
+                    .paint(&paint);
+                root_canvas.save_layer(&save_layer_rec);
+                root_canvas.restore();
+            }
+        }
+
+        let paint = Paint::default()
+            .set_anti_alias(false)
+            .set_color(Color::from_argb(255, 255, 255, default_background.a()))
+            .set_blend_mode(BlendMode::SrcOver)
+            .to_owned();
+
+        let save_layer_rec = SaveLayerRec::default().bounds(&bound_rect).paint(&paint);
+
+        root_canvas.save_layer(&save_layer_rec);
+        root_canvas.clear(default_background.with_a(255));
+
+        let regions = self
+            .windows
+            .iter()
+            .map(|window| window.pixel_region(font_dimensions))
+            .collect::<Vec<_>>();
+
+        let blend = self.uniform_background_blend();
+
+        self.windows.iter_mut().for_each(|window| {
+            window.update_blend(blend);
+        });
+
+        let mut ret = vec![];
+
+        (0..self.windows.len()).for_each(|i| {
+            let window = &mut self.windows[i];
+            window.draw_background_surface(root_canvas, &regions[i], font_dimensions);
+        });
+        (0..self.windows.len()).for_each(|i| {
+            let window = &mut self.windows[i];
+            window.draw_foreground_surface(root_canvas, &regions[i], font_dimensions);
+
+            ret.push(WindowDrawDetails {
+                id: window.id,
+                region: regions[i],
+                floating_order: window.anchor_info.as_ref().map(|v| v.sort_order),
+            });
+        });
+
+        root_canvas.restore();
+
+        root_canvas.restore();
+
+        ret
+    }
+
+    pub fn uniform_background_blend(&self) -> u8 {
+        self.windows
+            .iter()
+            .filter_map(|window| window.get_smallest_blend_value())
+            .min()
+            .unwrap_or(0)
+    }
+
+    fn _draw_shadow(&self, root_canvas: &Canvas, path: &Path, settings: &RendererSettings) {
```

Comment:
nit: Why is an underscore prefix used here? If its to indicate that this function is internal, I think it should be removed since it doesn't match the style for the rest of the project (barring the profiling code where this pattern has snuck in)

Reply:
BTW, for the profiling code, there's a reason. The `_LocationData` and `_Zone` have to be public but are only supposed to be used indirectly through the macros.